# Исследование рынка видеоигр

- Автор: Васильев Арсений
- Дата: 26.05.2026

### Цели и задачи проекта

**Цель проекта**

Провести комплексную предобработку, фильтрацию и категоризацию исторических данных о продажах и оценках видеоигр за период с 2000 по 2013 год включительно для подготовки качественного и корректного датасета к последующему анализу.

**Задачи проекта**

Первичное знакомство с данными:
- загрузить датасет и изучить его общую структуру;
- оценить соответствие данных их описанию и выявить первичные проблемные зоны.

Предобработка данных:
- привести наименования всех столбцов к единому стандартному стилю;
- обработать некорректные типы данных;
- проанализировать характер и причины пропусков, провести их обработку;
- устранить неявные и явные дубликаты;
- рассчитать объем и долю удаленных в ходе очистки данных.

Фильтрация данных:
- отсечь лишние временные периоды и сформировать отдельный актуальный срез данных за 2000–2013 годы включительно.

Категоризация данных:
- разделить игры по оценкам пользователей на три группы: высокая (8–10), средняя (3–8) и низкая (0–3);
- разделить игры по оценкам критиков на три группы: высокая (80–100), средняя (30–80) и низкая (0–30);
- определить топ-7 игровых платформ по количеству выпущенных игр за выбранный период.

Формулирование выводов:
- написать общий итог проделанной работы и дать подробное описание получившегося.

### Описание данных

В качестве исходных данных используется файл /datasets/new_games.csv, содержащий информацию о продажах компьютерных игр, их жанрах, платформах, а также оценках пользователей, экспертов и возрастных рейтингах.

Датасет включает в себя следующие признаки (после приведения к стандарту snake_case):
- name - название игры;
- platform - название игровой платформы;
- year_of_release - год выпуска игры;
- genre - жанр игры;
- na_sales - объём продаж в Северной Америке (в миллионах проданных копий);
- eu_sales - объём продаж в Европе (в миллионах проданных копий);
- jp_sales - объём продаж в Японии (в миллионах проданных копий);
- other_sales - объём продаж в других странах (в миллионах проданных копий);
- critic_score - оценка критиков и экспертов (по 100-балльной шкале, от 0 до 100);
- user_score - оценка пользователей (по 10-балльной шкале, от 0 до 10);
- rating - возрастной рейтинг, присвоенный организацией ESRB (Entertainment Software Rating Board), определяющий подходящую возрастную категорию для игры.

### Содержимое проекта

Шаг 1. Загрузка и знакомство с данными
- 1.1. Импорт библиотек и загрузка датасета.
- 1.2. Первичный визуальный анализ и вывод общей информации.
- 1.3. Вывод о качестве, объёме и структуре предоставленных данных.

Шаг 2. Проверка ошибок в данных и их предобработка
- 2.1. Изменение стиля наименования столбцов.
- 2.2. Анализ и преобразование типов данных.
- 2.3. Выявление, анализ причин и обработка пропущенных значений.
- 2.4. Поиск и устранение явных и неявных дубликатов.
- 2.5. Оценка масштаба изменений и общий промежуточный вывод по предобработке.

Шаг 3. Фильтрация данных
- 3.1. Создание актуального среза данных за период 2000–2013 гг.

Шаг 4. Категоризация данных
- 4.1. Категоризация игр по оценкам пользователей.
- 4.2. Категоризация игр по оценкам критиков.
- 4.3. Выделение топ-7 игровых платформ по количеству релизов.

Шаг 5. Общий вывод
- 5.1. Резюме проделанной работы, описание финального датасета и созданных признаков.

---

## Загрузка данных и знакомство с ними

Загрузим библиотеку Pandas и данные датасета `/datasets/new_games.csv`:

In [1]:
import pandas as pd

df = pd.read_csv('/datasets/new_games.csv')

Ознакомимся с данными с помощью метода info():


In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


Посмотрим первые строки датасета с помощью метода head():

In [3]:
df.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


Объем и структура:
- предоставленный датасет содержит 16 956 строк и 11 столбцов;
- содержимое полностью соответствует описанию (названия игр, платформы, годы выпуска, жанры, продажи по регионам, оценки критиков/пользователей и возрастной рейтинг ESRB).

Соответствие типов данных и аномалии:
- Year of Release имеет тип float64, его необходимо будет привести к целочисленному типу int;
- EU sales и JP sales имеют тип object (строковый), в то время как продажи в Северной Америке (NA sales) и других странах (Other sales) распознались как float64, значит в столбцах европейских и японских продаж присутствуют текстовые артефакты, которые мешают корректному анализу;
- User Score имеет тип object. Столбец требует перевода в числовой формат.

Пропуски в данных:
- в столбцах Name и Genre пропущено по 2 значения;
- в столбце Year of Release пропущено 275 значений;
- аибольшее количество пропусков зафиксировано в оценках и рейтингах: в Critic Score (пропущено 8 714 значений), User Score (пропущено 6 804 значения) и Rating (пропущено 6 871 значение).

Стиль наименования столбцов:
- названия столбцов нарушают стиль snake_case, они написаны с заглавных букв и содержат пробелы. Необходимо перевести в нижний регистр и заменить пробелы на нижнее подчёркивание.

---

## Проверка ошибок в данных и их предобработка


### Названия, или метки, столбцов датафрейма

Выведем названия всех столбцов датафрейма и проверим их стиль написания:

In [4]:
# названия столбцов до изменения
df.columns.tolist()

['Name',
 'Platform',
 'Year of Release',
 'Genre',
 'NA sales',
 'EU sales',
 'JP sales',
 'Other sales',
 'Critic Score',
 'User Score',
 'Rating']

Приведем все столбцы к стилю snake_case (названия в нижнем регистре, вместо пробелов - подчёркивания):

In [5]:
# приводим к нижнему регистру
df.columns = df.columns.str.lower()
# заменяем пробелы на нижние подчеркивания
df.columns = df.columns.str.replace(' ', '_')

Проверим названия столбцов после приведения к snake_case:

In [6]:
# названия столбцов после изменения
df.columns.tolist()

['name',
 'platform',
 'year_of_release',
 'genre',
 'na_sales',
 'eu_sales',
 'jp_sales',
 'other_sales',
 'critic_score',
 'user_score',
 'rating']

### Типы данных

В столбце user_score найдены значения tbd. Фактически это пропущенное значение, которое, скорее всего, означает нехватку голосов для определения рейтинга, поэтому они были переведены в NaN:

In [7]:
df['user_score'] = pd.to_numeric(df['user_score'], errors='coerce')

Столбцы eu_sales и jp_sales содержат текстовое значение unknown. Это техническая ошибка или отсутствие данных при выгрузке, из-за чего Pandas считывал весь столбец, как строковый (object). После принудительного преобразования они стали полноценными вещественными числами (float64).

In [8]:
df['eu_sales'] = pd.to_numeric(df['eu_sales'], errors='coerce')
df['jp_sales'] = pd.to_numeric(df['jp_sales'], errors='coerce')

Проверим типы данных после преобразований:

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16950 non-null  float64
 6   jp_sales         16952 non-null  float64
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       7688 non-null   float64
 10  rating           10085 non-null  object 
dtypes: float64(7), object(4)
memory usage: 1.4+ MB


### Наличие пропусков в данных

Посчитаем количество пропусков в каждом столбце в абсолютных и относительных значениях:

In [10]:
absolute = df.isna().sum()
relative = round(df.isna().mean() * 100, 2)

print('Количество пропусков:\n', absolute)
print('Доля пропусков:\n', relative)

Количество пропусков:
 name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64
Доля пропусков:
 name                0.01
platform            0.00
year_of_release     1.62
genre               0.01
na_sales            0.00
eu_sales            0.04
jp_sales            0.02
other_sales         0.00
critic_score       51.39
user_score         54.66
rating             40.52
dtype: float64


name и genre - пропущено всего по 2 значения (0.01%). Это случайные единичные пропуски, их можно удалить:

In [11]:
# удаляем строки где пропущены имя игры или жанр
df = df.dropna(subset=['name', 'genre'])

year_of_release - пропущено 275 значений (1.62%). У некоторых старых или редких игр год не был внесен в базу. Эти строки также можно удалить, тк наш дальнейший анализ привязан к временному отрезку 2000–2013 гг. Также преобразуем год выхода в целые числа int:

In [12]:
# удаляем строки где пропущен год выхода
df = df.dropna(subset=['year_of_release'])

# переводим года выпуска в обычные целые числа int
df['year_of_release'] = df['year_of_release'].astype(int)

eu_sales (6 пропусков) и jp_sales (4 пропуска) - возникли после очистки unknown. Можно заполнить средними значениями в зависимости от платформы и года выпуска:

In [13]:
# ЗАПОЛНЯЕМ ПРОПУСКИ В ЕВРОПЕЙСКИХ ПРОДАЖАХ

# получаем список строк с пропусками в столбце eu_sales
eu_rows = df[df['eu_sales'].isna()].index

for i in eu_rows:
    # получаем платформу и год для строки с пропуском
    platform = df.loc[i, 'platform']
    year = df.loc[i, 'year_of_release']
    
    # выбираем игры с такой же платформой и годом
    games = df[(df['platform'] == platform) & (df['year_of_release'] == year)]
    
    # вычисляем среднее значение продаж
    mean = games['eu_sales'].mean()
    
    # заменяем пропуски на среднее значение
    df.loc[i, 'eu_sales'] = mean
    
# ЗАПОЛНЯЕМ ПРОПУСКИ В ЯПОНСКИХ ПРОДАЖАХ

eu_rows = df[df['jp_sales'].isna()].index

for i in eu_rows:
    platform = df.loc[i, 'platform']
    year = df.loc[i, 'year_of_release']
    
    games = df[(df['platform'] == platform) & (df['year_of_release'] == year)]
    
    mean = games['jp_sales'].mean()
    
    df.loc[i, 'jp_sales'] = mean

critic_score (51.39%) и user_score (54.66%) - пропущено более половины данных. Заполнять их искусственно (средним или медианой) нельзя, это исказит распределение оценок. Оставим пропуски как есть (NaN), чтобы не вносить погрешность.

rating - пропущено 40.52% данных. Некоторые игры не получили рейтинг от американской копмании ESRB. Заменим пропуски на строковое значение-индикатор unknown:

In [14]:
# заполняем пропущенный возрастной рейтинг индикатором unknown
df['rating'] = df['rating'].fillna('unknown')

**Возможные причины возникновения пропусков**
1. name и genre - MCAR (полностью случайные пропуски). Эти пропуски единичны (всего по 2 значения на всю таблицу). Они возникли изза случайной технической ошибки при копировании данных или человеческого фактора (оператор случайно пропустил поле при ручном вводе).
2. year_of_release - MCAR / MAR (полностью случайные или случайные). Пропуски составляют 1.62%. Часть из них - технические ошибки при переносе или слиянии баз данных разных игровых платформ (MCAR). Другая часть связана с тем, что для старых, редких или малоизвестных игр точный год официального релиза мог быть не зафиксирован на момент составления таблицы (MAR).
3. eu_sales и jp_sales - MCAR (полностью случайные пропуски). Образовались в результате предобработки, когда мы принудительно заменили некорректные текстовые заглушки unknown на числовой формат NaN. Появление unknown в исходном файле вызвано отсутствием информации о продажах на момент выгрузки данных.
4. critic_score и user_score - MAR / MNAR (случайные или неслучайные). Пропущено более половины всех данных (51.39% и 54.66%). У большого количества ретро-игр ранее не существовало оценок в цифровом виде. Некоторые малобюджетные игры могут вообще не получать обзоров от СМИ и критиков. В случае с пользователями часть пропусков - это бывшие текстовые значения tbd (будет определено). Рейтинг у этих игр отсутствует физически, так как пользователи оставили мало отзывов, и система не смогла рассчитать средний балл.
5. rating - MNAR (неслучайные пропуски). Пропущено 40.52% данных. Это связано со спецификой признака. Организация ESRB, присваивающая данный рейтинг, является американской. Игры, выпущенные до её основания (1994 год) или издававшиеся исключительно для внутренних рынков Японии и Европы, этот рейтинг не получали.

### Явные и неявные дубликаты в данных

Изучим уникальные значение в категориальных данных:

In [15]:
# проверка уникальных значений категориальных полей
for column in ['genre','platform','rating','year_of_release']:
    print(f'Уникальные значения в столбце {column}:')
    print(df[column].sort_values().unique())
    print()

Уникальные значения в столбце genre:
['ACTION' 'ADVENTURE' 'Action' 'Adventure' 'FIGHTING' 'Fighting' 'MISC'
 'Misc' 'PLATFORM' 'PUZZLE' 'Platform' 'Puzzle' 'RACING' 'ROLE-PLAYING'
 'Racing' 'Role-Playing' 'SHOOTER' 'SIMULATION' 'SPORTS' 'STRATEGY'
 'Shooter' 'Simulation' 'Sports' 'Strategy']

Уникальные значения в столбце platform:
['2600' '3DO' '3DS' 'DC' 'DS' 'GB' 'GBA' 'GC' 'GEN' 'GG' 'N64' 'NES' 'NG'
 'PC' 'PCFX' 'PS' 'PS2' 'PS3' 'PS4' 'PSP' 'PSV' 'SAT' 'SCD' 'SNES' 'TG16'
 'WS' 'Wii' 'WiiU' 'X360' 'XB' 'XOne']

Уникальные значения в столбце rating:
['AO' 'E' 'E10+' 'EC' 'K-A' 'M' 'RP' 'T' 'unknown']

Уникальные значения в столбце year_of_release:
[1980 1981 1982 1983 1984 1985 1986 1987 1988 1989 1990 1991 1992 1993
 1994 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004 2005 2006 2007
 2008 2009 2010 2011 2012 2013 2014 2015 2016]




<div class="alert alert-block alert-info">
   
<b>Совет:</b> Лаконичнее делать осмотр через цикл
    
    for column in ['platform','year_of_release','genre','rating']:
        print(f'Уникальные значения в столбце {column}:')
        print(df[column].sort_values().unique())
        print()

</div>



В жанрах (genre) обнаружены явные проблемы с регистром. Один и тот же жанр записан по-разному, например, 'Sports' и 'SPORTS', 'Action' и 'ACTION'. Необходимо привести все значения к нижнему регистру:

In [16]:
df['genre'] = df['genre'].str.lower()

В платформах (platform) все названия уникальны, записаны корректно, аномалий или повторов не обнаружено.

В рейтингах (rating) все значения стандартны для системы ESRB. Приведем все значения к верхнему регистру для единообразия:

In [17]:
df['rating'] = df['rating'].str.upper()

В годах выпуска (year_of_release) все значения представляют собой корректные целые числа, аномальных годов нет.

Теперь подсчитаем и удалим явные дубликаты:

In [18]:
duplicates_count = df.duplicated().sum()
print('Количество явных дубликатов до удаления:', duplicates_count)
df = df.drop_duplicates()

Количество явных дубликатов до удаления: 235


**Промежуточный вывод по дубликатам:**
- В ходе анализа категориальных признаков были обнаружены неявные дубликаты в столбце genre (названия одних и тех же жанров были записаны в разном регистре). 
- После приведения столбца genre к нижнему регистру, а столбца rating к верхнему, в данных было обнаружено и успешно удалено 235 строк (явные дубликаты). На текущий момент дубликаты в датасете отсутствуют.

Подсчет удалённых строк:

In [19]:
# исходное количество строк
rows = 16956 

# текущее количество строк
final_rows = len(df) 

# вычисляем разницу
deleted_rows = rows - final_rows
deleted_percent = (deleted_rows / rows) * 100

print(f'Исходное количество строк: {rows}')
print(f'Количество строк после очистки: {final_rows}')
print(f'Абсолютное количество удаленных строк: {deleted_rows}')
print(f'Относительное количество удаленных строк: {deleted_percent:.2f}%')

Исходное количество строк: 16956
Количество строк после очистки: 16444
Абсолютное количество удаленных строк: 512
Относительное количество удаленных строк: 3.02%


**Общий промежуточный вывод**

В процессе предобработки исходного датасета были выполнены следующие шаги:
1. Коррекция заголовков:
    - названия всех колонок приведены к единому стандартному стилю snake_case.
2. Обработка аномалий и типов данных: 
    - из столбцов с продажами (eu_sales, jp_sales) и оценками пользователей (user_score) удалены строковые артефакты (unknown, tbd), которые мешали математическим расчетам. Столбцы переведены в числовые форматы;
    - столбец с годами выпуска year_of_release преобразован в целочисленный тип int.
3. Заполнение и удаление пропусков:
    - пропуски в продажах по Европе и Японии были заполнены средними значениями продаж игр для этой же платформы в этот же год выпуска;
    - пропуски в столбце возрастного рейтинга rating заполнены текстовым индикатором unknown, чтобы сохранить данные для анализа и выделить игры без рейтинга ESRB в отдельную группу;
    - пропуски в оценках критиков и пользователей оставлены без изменений, чтобы не искажать реальную статистику;
    - строки с пропущенными ключевыми значениями (имя, жанр, год выпуска) были удалены.
4. Устранение дубликатов:
    - устранены неявные дубликаты в жанрах за счет исправления регистра, после чего из таблицы удален 235 строк.

**Итог:** Суммарно за весь этап предобработки было удалено 512 строк, что составляет 3.02% от исходного объема данных.

---

## Фильтрация данных

На основе очищенных данных сформируем новый актуальных срез df_actual, в который войдет временной отрезок с 2000 по 2013 год включительно:

In [20]:
# фильтруем данные
df_actual = df[(df['year_of_release'] >= 2000) & (df['year_of_release'] <= 2013)].copy()

# проверка фильтрации
print("Минимальный год:", df_actual['year_of_release'].min())
print("Максимальный год:", df_actual['year_of_release'].max())

# подсчитаем количество строк после фильтрации
print("Количество строк в актуальном срезе:", len(df_actual))

Минимальный год: 2000
Максимальный год: 2013
Количество строк в актуальном срезе: 12781


---

## Категоризация данных

Проведем категоризацию по оценкам пользователей (user_score), разобъем все игры на категории:
- высокая оценка - от 8 до 10 включительно;
- средняя оценка - от 3 до 8, не включая правую границу интервала;
- низкая оценка - от 0 до 3, не включая правую границу интервала.

In [21]:
df_actual['user_category'] = pd.cut(
    df_actual['user_score'],
    bins = [0, 3, 8, 10.01],
    labels = ['низкая оценка', 'средняя оценка', 'высокая оценка'],
    right = False
)

Теперь проведем категоризацию по оценкам критиков (critic_score), выделим такие категории:
- высокая оценка - от 80 до 100 включительно;
- средняя оценка - от 30 до 80, не включая правую границу интервала;
- низкая оценка - от 0 до 30, не включая правую границу интервала).

In [22]:
df_actual['critic_category'] = pd.cut(
    df_actual['critic_score'],
    bins = [0, 30, 80, 100.01],
    labels = ['низкая оценка', 'средняя оценка', 'высокая оценка'],
    right = False
)

Проверим результат - сгруппируем данные по выделенным категориям и посчитаем количество игр в каждой категории:

In [23]:
user = df_actual.groupby('user_category')['name'].count()
print('Распределение категорий по оценкам пользователей:\n', user)

critic = df_actual.groupby('critic_category')['name'].count()
print('Распределение категорий по оценкам критиков:\n', critic)

Распределение категорий по оценкам пользователей:
 user_category
низкая оценка      116
средняя оценка    4081
высокая оценка    2286
Name: name, dtype: int64
Распределение категорий по оценкам критиков:
 critic_category
низкая оценка       55
средняя оценка    5422
высокая оценка    1692
Name: name, dtype: int64


Также определим 7 самых популярных игровых платформ по количеству игр, выпущенных за весь актуальный период:

In [24]:
top7_platforms = df_actual.groupby(['platform'])['name'].count().sort_values(ascending=False).head(7)

print('Топ-7 платформ по количеству игр, выпущенных в 2000-2013 гг.:\n', top7_platforms)

Топ-7 платформ по количеству игр, выпущенных в 2000-2013 гг.:
 platform
PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1087
GBA      811
Name: name, dtype: int64


---

## Итоговый вывод

**Проделанная работа:**
- Очистка данных: привели заголовки к единому стилю snake_case, удалили текстовые артефакты (tbd, unknown) и исправили типы данных.
- Пропуски и дубликаты: заполнили пустые значения в продажах средними показателями по группам, а в рейтинге - индикатором unknown. Избавились от неявных повторов в жанрах и удалили 235 явных дубликатов.
- Потери: общая доля удалённых строк составила 3.02% (512 строк).

**Описание актуального среза данных (df_actual):**
- Критерий отбора: в датасет вошли игры, выпущенные в период с 2000 по 2013 год включительно.
- Итоговый объём: 12 781 строк.

**Описание новых полей:**
- user_category - оценка игры пользователями: низкая (от 0 до 3), средняя (от 3 до 8), высокая (от 8 до 10 включительно).
- critic_category - оценка игры критиками: низкая (от 0 до 30), средняя (от 30 до 80), высокая (от 80 до 100 включительно).